src/main/dx.ts

```diff
@@ -0,0 +1,423 @@
+  ) {
+    this.client = new Client(token);
+    this.src = src;
+    this.size = fs.statSync(this.src).size;
+    this.progressCb = progressCb;
+    this.finishedCb = finishedCb;
+  }
+
+  public async prepare(projectId: string): Promise<utils.IByteRange[]> {
+    const { maximumPartSize } = await fileUploadParameters(
+      this.client,
+      projectId,
+    );
+
+    return utils.byteRanges(this.size, maximumPartSize);
+  }
+
+  public async start(projectId: string, dst: string) {
+    const ranges = await this.prepare(projectId);
+    const name = path.basename(this.src);
+
+    const { id } = await this.client.file.new({
+      folder: dst,
+      name,
+      parents: true,
+      project: projectId,
+      // tags: [config.NEEDS_ANALYSIS_TAG],
+    });
+
+    for (let i = 0; i < ranges.length; i++) {
+      const { start, end } = ranges[i];
+      await this.transfer(id, i + 1, start, end);
+    }
+
+    await this.client.file.close(id);
+
+    this.finishedCb(null, {});
+  }
+
+  public abort() {
+    if (request) {
+      request.abort();
+    }
+  }
+
+  private async transfer(
+    id: string,
+    i: number,
+    start: number,
+    end: number,
+  ): Promise<{}> {
+    const reader = fs.createReadStream(this.src, { start, end });
+
+    const size = end - start + 1;
+    const md5 = await utils.md5Sum(this.src, start, end);
+
+    const { url, headers } = await this.client.file.upload(id, {
+      index: i,
+      md5,
+      size,
+    });
+
+    return new Promise(resolve => {
+      this.request = request(
+        url,
+        { body: reader, headers, method: 'PUT' },
+        async (error: any, response: any) => {
+          if (error) {
+            this.finishedCb(error, null);
+          } else {
+            this.bytesRead += size;
+            resolve();
+          }
+        },
+      );
+
+      this.request.on('drain', () => {
+        const r: any = this.request;
+        const { bytesWritten } = r.req.connection;
+        const percent = (this.bytesRead + bytesWritten) / this.size * 100;
+        this.progressCb(percent);
+      });
+
+      this.request.on('abort', () => {
+        this.finishedCb(new Error('upload aborted'), null);
+      });
+    });
+  }
+}
+
+/**
+ * Uploads a file to a DNAnexus project via the dx command line utility.
+ *
+ * @param token {string} Token to identify ourselves to the API.
+ * @param projectId {string} DNAnexus ID of projectId being uploaded to.
+ * @param file {IRemoteLocalFilePairForUpload} Local and remote file name pair.
+ * @param progressCb {ResultCallback} Periodically called with updates on the
+ *                                    upload progress.
+ * @param finishedCb {SuccessCallback} Called on successful upload or failure.
+ * @returns {UploadTransfer} The upload transfer request.
+ */
+export function uploadFile(
+  token: string,
+  file: IRemoteLocalFilePairForUpload,
+  progressCb: ResultCallback<number>,
+  finishedCb: SuccessCallback,
```

Comment:
@zaeleus I changed the callbacks to include some generics, but I couldn't discern what the correct type for this would be. Can you add?